File: app/src/main/java/email/schaal/ocreader/view/AddNewFeedDialogFragment.java
package email.schaal.ocreader.view;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;

import email.schaal.ocreader.ManageFeedsActivity;
import email.schaal.ocreader.R;
import email.schaal.ocreader.database.model.Feed;
import email.schaal.ocreader.databinding.FragmentAddNewFeedBinding;

/**
 * Display form to add new Feed
 */
public class AddNewFeedDialogFragment extends DialogFragment {
    public static final String ARG_URL = "url";
    public static final String ARG_FOLDER_ID = "folder_id";
    public static final String ARG_FINISH_AFTER_CLOSE = "finish_after_close";
    private static final String ARG_FEED_ID = "feed_id";

    private FeedManageListener listener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
           listener = (FeedManageListener) context;
        } catch(ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement FeedManageListener");
        }
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        ManageFeedsActivity activity = (ManageFeedsActivity) getActivity();

        AlertDialog.Builder builder = new AlertDialog.Builder(activity);

        final FragmentAddNewFeedBinding binding = FragmentAddNewFeedBinding.inflate(activity.getLayoutInflater());

        final Bundle arguments = getArguments();
        final long feedId = arguments.getLong(ARG_FEED_ID, -1);
        final boolean newFeed = feedId < 0;

        builder.setTitle(newFeed ? R.string.add_new_feed : R.string.edit_feed);

        binding.feedUrl.setEnabled(newFeed);

        final boolean finishAfterClose;

        binding.folder.setAdapter(activity.getFolderSpinnerAdapter());

        binding.feedUrl.setText(arguments.getString(ARG_URL));
        binding.folder.setSelection(activity.getFolderSpinnerAdapter().getPosition(arguments.getLong(ARG_FOLDER_ID, 0)));
        finishAfterClose = arguments.getBoolean(ARG_FINISH_AFTER_CLOSE, false);

        builder.setPositiveButton(newFeed ? R.string.add : R.string.save, (dialogInterface, i) -> {
            if(listener != null) {
                if(newFeed)
                    listener.addNewFeed(binding.feedUrl.getText().toString(), binding.folder.getSelectedItemId(), finishAfterClose);
                else
                    listener.changeFeed(feedId, binding.folder.getSelectedItemId());
            }
        });

        builder.setView(binding.getRoot());

        return builder.create();
    }

    /**
     * Show feed edit dialog
     * @param activity Activity to get the FragmentManager
     * @param feed feed to edit or add (id < 0 means add new feed, id >=0 means edit existing feed)
     * @param finishAfterClose should the activity be finished after operation was successful
     */
    public static void show(Activity activity, @Nullable Feed feed, boolean finishAfterClose) {
        AddNewFeedDialogFragment dialogFragment = new AddNewFeedDialogFragment();

        Bundle bundle = new Bundle();

        bundle.putBoolean(ARG_FINISH_AFTER_CLOSE, finishAfterClose);

        if(feed != null) {
            bundle.putString(ARG_URL, feed.getUrl());
            if (feed.getId() >= 0) {
                bundle.putLong(ARG_FEED_ID, feed.getId());
                bundle.putLong(ARG_FOLDER_ID, feed.getFolderId());
            }
        }

        dialogFragment.setArguments(bundle);
        dialogFragment.show(activity.getFragmentManager(), "newfeed");
    }
}

File: app/src/main/java/email/schaal/ocreader/view/AddNewFeedDialogFragment.kt
package email.schaal.ocreader.view

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import email.schaal.ocreader.ManageFeedsActivity
import email.schaal.ocreader.R
import email.schaal.ocreader.database.model.Feed
import email.schaal.ocreader.databinding.FragmentAddNewFeedBinding

/**
 * Display form to add new Feed
 */
class AddNewFeedDialogFragment : BottomSheetDialogFragment() {
    private var listener: FeedManageListener? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = try {
            context as FeedManageListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement FeedManageListener")
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val activity = requireActivity() as ManageFeedsActivity

        val arguments = requireArguments()
        val feedId = arguments.getLong(ARG_FEED_ID, -1)
        val newFeed = feedId < 0

        val binding = FragmentAddNewFeedBinding.inflate(activity.layoutInflater).apply {
            feedUrl.isEnabled = newFeed
            folder.adapter = activity.folderSpinnerAdapter
            feedUrl.setText(arguments.getString(ARG_URL))
            folder.setSelection(activity.folderSpinnerAdapter.getPosition(arguments.getLong(ARG_FOLDER_ID, 0)))
        }

        return AlertDialog.Builder(activity).setTitle(if (newFeed) R.string.add_new_feed else R.string.edit_feed)
                .setPositiveButton(if (newFeed) R.string.add else R.string.save) { _, _ ->
                    if (newFeed)
                        listener?.addNewFeed(binding.feedUrl.text.toString(), binding.folder.selectedItemId, arguments.getBoolean(ARG_FINISH_AFTER_CLOSE, false))
                    else
                        listener?.changeFeed(feedId, binding.folder.selectedItemId)
                }
                .setView(binding.root)
                .create()
    }

    companion object {
        const val ARG_URL = "url"
        const val ARG_FOLDER_ID = "folder_id"
        const val ARG_FINISH_AFTER_CLOSE = "finish_after_close"
        private const val ARG_FEED_ID = "feed_id"
        /**
         * Show feed edit dialog
         * @param activity Activity to get the FragmentManager
         * @param feed feed to edit or add (id < 0 means add new feed, id >=0 means edit existing feed)
         * @param finishAfterClose should the activity be finished after operation was successful
         */
        fun show(activity: FragmentActivity, feed: Feed?, finishAfterClose: Boolean) {
            val dialogFragment = AddNewFeedDialogFragment()

            dialogFragment.arguments = Bundle().apply {
                putBoolean(ARG_FINISH_AFTER_CLOSE, finishAfterClose)
                feed?.let { feed ->
                    putString(ARG_URL, feed.url)
                    if (feed.id >= 0) {
                        putLong(ARG_FEED_ID, feed.id)
                        putLong(ARG_FOLDER_ID, feed.folderId)
                    }
                }
            }
            dialogFragment.show(activity.supportFragmentManager, "newfeed")
        }
    }
}